benchmarks/babelstream/src/babelStreamMainTest.cpp

```diff
@@ -312,136 +345,252 @@ void testKernels()
+            sumErrB += std::fabs(bufHostOutputB[static_cast<Idx>(i)] - expectedB);
+            sumErrA += std::fabs(bufHostOutputA[static_cast<Idx>(i)] - expectedA);
+        }
 
-        alpaka::memcpy(queue, bufHostSumPerBlock, bufAccSumPerBlock, gridBlockExtent);
+        // Normalize and compare sum of the errors
+        // Use a different equality check if floating point errors exceed precision of FuzzyEqual function
+        REQUIRE(FuzzyEqual(sumErrC / static_cast<DataType>(arraySize) / expectedC, static_cast<DataType>(0.0)));
+        REQUIRE(FuzzyEqual(sumErrB / static_cast<DataType>(arraySize) / expectedB, static_cast<DataType>(0.0)));
+        REQUIRE(FuzzyEqual(sumErrA / static_cast<DataType>(arraySize) / expectedA, static_cast<DataType>(0.0)));
         alpaka::wait(queue);
 
-        DataType const* sumPtr = std::data(bufHostSumPerBlock);
-        auto const result = std::reduce(sumPtr, sumPtr + gridBlockExtent, DataType{0});
-        // Since vector values are 1, dot product should be identical to arraySize
-        REQUIRE(FuzzyEqual(static_cast<DataType>(result), static_cast<DataType>(arraySize * 2)));
-        // Add workdiv to the list of workdivs to print later
-        metaData.setItem(BMInfoDataType::WorkDivDot, workDivDot);
-    }
+        // Test Dot kernel with specific blocksize which is larger than 1
+        if constexpr(alpaka::accMatchesTags<TAcc, alpaka::TagGpuCudaRt, alpaka::TagGpuHipRt, alpaka::TagGpuSyclIntel>)
+        {
+            using WorkDiv = alpaka::WorkDivMembers<Dim, Idx>;
+            // Threads per block for Dot kernel
+            constexpr Idx blockThreadExtent = blockThreadExtentMain;
+            // Blocks per grid for Dot kernel
+            const Idx gridBlockExtent = static_cast<Idx>(dotGridBlockExtent);
+            // Vector of sums of each block
+            auto bufAccSumPerBlock = alpaka::allocBuf<DataType, Idx>(devAcc, gridBlockExtent);
+            auto bufHostSumPerBlock = alpaka::allocBuf<DataType, Idx>(devHost, gridBlockExtent);
+            // A specific work-division is used for dotKernel
+            auto const workDivDot = WorkDiv{Vec{gridBlockExtent}, Vec{blockThreadExtent}, Vec::all(1)};
+
+            measureKernelExec(
+                [&]()
+                {
+                    alpaka::exec<Acc>(
+                        queue,
+                        workDivDot,
+                        DotKernel(), // Dot kernel
+                        bufAccInputAPtr,
+                        bufAccInputBPtr,
+                        alpaka::getPtrNative(bufAccSumPerBlock),
+                        static_cast<alpaka::Idx<Acc>>(arraySize));
+                },
+                "DotKernel");
+
+            alpaka::memcpy(queue, bufHostSumPerBlock, bufAccSumPerBlock, gridBlockExtent);
+            alpaka::wait(queue);
 
+            DataType const* sumPtr = std::data(bufHostSumPerBlock);
+            float const result = std::reduce(sumPtr, sumPtr + gridBlockExtent, 0.0f);
```

Comment:
This and the memcpy has be be part of the measurement, because that's our fault that we not execute the full reduction on device. 
To have a fair comparison the allocation of the result must be part of `measureKernelExec` too but the cuda upstream implementation is cheating here too so I assume this is fine to be allocated outside of `measureKernelExec`

Reply:
In [the original code](https://github.com/UoB-HPC/BabelStream/blob/2f00dfb7f8b7cfe8c53d20d5c770bccbf8673440/src/cuda/CUDAStream.cu), they don't do full reduction. Kernel calculates dot product at the block level but does not sum.


```
template <class T>
__global__ void dot_kernel(const T * a, const T * b, T * sum, int array_size)
{
  __shared__ T tb_sum[TBSIZE];
  int i = blockDim.x * blockIdx.x + threadIdx.x;
  const size_t local_i = threadIdx.x;
  tb_sum[local_i] = {};
  for (; i < array_size; i += blockDim.x*gridDim.x)
    tb_sum[local_i] += a[i] * b[i];

  for (int offset = blockDim.x / 2; offset > 0; offset /= 2)
  { __syncthreads();
     if (local_i < offset)
     { tb_sum[local_i] += tb_sum[local_i+offset];
     }
  }
  if (local_i == 0)
    sum[blockIdx.x] = tb_sum[local_i];
}
```

Reply:
ok, I saw these are also part of measurement. Thanks.

Reply:
done, thanks.